### PR TITLE
Add Tracer

### PR DESCRIPTION
**Pull request by DavidArchibald**
I decided against making this PR a few months ago once I no longer found it necessary but this allows us to add in Open Telemetry more easily so I think it's worth it now.

### CODE REVIEW

**Comment by DavidArchibald** [comment]
Changed from a rewriter to a tracer

File: sling.go [hunk 1]
```diff
@@ -432,11 +437,43 @@ func (s *Sling) Receive(successV, failureV interface{}) (*http.Response, error)
...
 func (s *Sling) ReceiveWithContext(ctx context.Context, successV, failureV interface{}) (*http.Response, error) {
-	req, err := s.RequestWithContext(ctx)
+	req, err := s.requestWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return s.doDecode(req, successV, failureV)
+}
+
+func (s *Sling) Do(ctx context.Context) (*http.Response, error) {
```
Comment (robarchibald) [inline sling.go:447]: I don't think this makes sense. I think you should have to pass in the request. It should look just like the `http` package signature IMHO

File: tracer.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	"context"
+	"io"
+)
+
+type Tracer interface {
+	BeginTrace(ctx context.Context)
+	EndTrace(ctx context.Context)
+}
+
+type bodyWithTracer struct {
```
Comment (robarchibald) [inline tracer.go:14]: I don't see any tests for this new code

**Review by robarchibald — CHANGES_REQUESTED** [review]
Looks good. Just a few comments

File: sling.go (hunk 1 above)
Comment (DavidArchibald) [reply sling.go:447]: Talked about this in person and decided to leave it as is.

**Review by robarchibald — APPROVED** [review]
Thanks!

**Pull request merged**